chasm/tree.go

```diff
@@ -1260,31 +1273,96 @@ func (n *Node) closeTransactionHandleRootLifecycleChange() (bool, error) {
 }
 
 func (n *Node) closeTransactionForceUpdateVisibility(
-	nextVersionedTransition *persistencespb.VersionedTransition,
+	rootLifecycleChanged bool,
 ) error {
-	for _, child := range n.children {
+	var visibilityNode *Node
+	var visibilityPath []string
+	for childName, child := range n.children {
 		componentAttr := child.serializedNode.Metadata.GetComponentAttributes()
-		if componentAttr == nil ||
-			componentAttr.Type != visibilityComponentFqType ||
-			transitionhistory.Compare(child.serializedNode.Metadata.LastUpdateVersionedTransition, nextVersionedTransition) == 0 {
-			// Skip force update is the node is not a visibility component or already updated in this transition.
+		if componentAttr == nil {
 			continue
 		}
 
-		mutableContext := NewMutableContext(context.Background(), n)
-		visComponent, err := child.Component(mutableContext, ComponentRef{})
+		if child.valueState == valueStateNeedSerialize {
+			if rc, ok := n.registry.componentFor(child.value); ok && rc.fqType() == visibilityComponentFqType {
+				visibilityNode = child
+				visibilityPath = []string{childName}
+				break
+			}
+		} else if componentAttr.Type == visibilityComponentFqType {
+			visibilityNode = child
+			visibilityPath = []string{childName}
+			break
+		}
+
+	}
+	if visibilityNode == nil {
+		return nil
+	}
+
+	needUpdate := rootLifecycleChanged
+
+	mutableContext := NewMutableContext(context.Background(), n)
+	immutableContext := mutableContext.ToImmutable()
+
+	// NOTE: It is a HACK to access visibility component with immutable Context,
+	// but later invoke a method that requires MutableContext.
+	//
+	// But here we only know if visibility component needs to be updated or not AFTER
+	// invoking a mutable method on it.
+	// To prevent unnecessary update and replication of its state, we first access it with a immutable Context,
+	// then, at the end of this method, when we are sure it needs to be updated,
+	// access the visibility component again with a MutableContext so that the changes can be
+	// serialized and persisted.
+	//
+	// TODO: have a better story around checking for updates and skip unnecessary updates even if a
+	// component is accessed with a MutableContext, but no changes are actually made.
+	// (Not just state, need to check for if there is any new tasks too.)
+	visComponent, err := visibilityNode.Component(immutableContext, ComponentRef{})
+	if err != nil {
+		return err
+	}
+
+	visibility, ok := visComponent.(*Visibility)
+	if !ok {
+		return serviceerror.NewInternalf("expected visibility component for component with type %s, but got %T", visibilityComponentFqType, visComponent)
+	}
+
+	rootComponent, err := n.Component(immutableContext, ComponentRef{})
+	if err != nil {
+		return err
+	}
+
+	saProvider, ok := rootComponent.(VisibilitySearchAttributesProvider)
+	if ok {
+		updated, err := visibility.SetSearchAttributes(mutableContext, saProvider.SearchAttributes(immutableContext))
```

Comment:
I don't think this is the desirable approach, there's no need to duplicate the data in both the root and the visibility components.